backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceEventsHandler.java

```diff
@@ -0,0 +1,202 @@
+     * BacktraceMetrics instance will send data to Backtrace.
+     */
+    private int maxNumEvents = 350;
+
+    public void setMaxNumEvents(int maxNumEvents) {
+        this.maxNumEvents = maxNumEvents;
+    }
+
+    public int getMaxNumEvents() {
+        return this.maxNumEvents;
+    }
+
+    /**
+     * Time between retries if metrics submission fails
+     */
+    private final int timeBetweenRetriesMillis;
+
+    /**
+     * List of events in the event queue
+     */
+    ConcurrentLinkedDeque<Event> events = new ConcurrentLinkedDeque<Event>();
+
+    /**
+     * Number of retry attempts
+     */
+    protected int numRetries = 0;
+
+    /**
+     * Http client
+     */
+    private final Api api;
+
+    /**
+     * Submission url
+     */
+    private final String submissionUrl;
+
+    /**
+     * User provided custom attributes
+     */
+    protected final Map<String, Object> customAttributes;
+
+    /**
+     * The application context
+     */
+    protected final Context context;
+
+    /**
+     * @param context
+     * @param api
+     * @param backtraceHandlerThread
+     * @param urlPrefix
+     * @param settings
+     */
+
+    public BacktraceEventsHandler(Context context, Map<String, Object> customAttributes,
+                                  Api api, final BacktraceHandlerThread backtraceHandlerThread, String urlPrefix, BacktraceMetricsSettings settings)
+    {
+        // This should always have a nonnull looper because BacktraceHandlerThread starts in the
+        // constructor and getLooper blocks until the looper is ready if the thread is started
+        super(backtraceHandlerThread.getLooper());
+        if (backtraceHandlerThread.getLooper() == null) {
+            throw new NullPointerException("Expected nonnull looper, this should not happen");
+        }
+        this.context = context;
+        this.customAttributes = customAttributes;
+        this.backtraceHandlerThread = backtraceHandlerThread;
+        this.api = api;
+        this.submissionUrl = settings.getBaseUrl() + "/" + urlPrefix + "/submit?token=" + settings.getToken() + "&universe=" + settings.getUniverseName();
+        this.timeBetweenRetriesMillis = settings.getTimeBetweenRetriesMillis();
+
+        long timeIntervalMillis = settings.getTimeIntervalMillis();
+
+        if (timeIntervalMillis != 0) {
+            final BacktraceEventsHandler handler = this;
+            handler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    handler.send();
+                    handler.postDelayed(this, timeIntervalMillis);
+                }
+            }, timeIntervalMillis);
+        }
+    }
+
+    public abstract void sendStartupEvent(String eventName);
+
+    public void send() {
+        sendEvents(events, 0);
+    }
+
+    private void sendEvents(ConcurrentLinkedDeque<Event> events, int attempts) {
+        if (events == null) {
```

Comment:
why do we need `attempts` parameter here? I think we should pass it to BacktraceApi

Reply:
You are right, it's not even used here. I think we should instead fix `numRetries` logic in `BacktraceEventsHandler.java` as you pointed out earlier

Reply:
Removed this and fixed numRetries logic, thanks @konraddysput !